codegen/src/main/scala/overflowdb/codegen/Helpers.scala

```diff
@@ -123,42 +116,69 @@ object Helpers {
-      case Cardinality.ISeq => s"IndexedSeq[$tpe]"
+      case Property.Cardinality.ZeroOrOne => s"Option[$tpe]"
+      case Property.Cardinality.One(_)    => tpe
+      case Property.Cardinality.List      => s"Seq[$tpe]"
+      case Property.Cardinality.ISeq => s"IndexedSeq[$tpe]"
     }
   }
 
-  def propertyBasedFields(properties: Seq[Property]): String = {
-    properties.map { property =>
-      val publicName = camelCase(property.name)
-      val fieldName = s"_$publicName"
-      val (publicType, tpeForField, fieldAccessor, defaultValue) = {
-        val valueType = typeFor(property.valueType)
-        getHigherType(property.cardinality) match {
-          case HigherValueType.None   => (valueType, valueType, fieldName, "null")
-          case HigherValueType.Option => (s"Option[$valueType]", valueType, s"Option($fieldName)", "null")
-          case HigherValueType.List   => (s"Seq[$valueType]", s"Seq[$valueType]", fieldName, "Nil")
-          case HigherValueType.ISeq   => (s"IndexedSeq[$valueType]", s"IndexedSeq[$valueType]", fieldName, "IndexedSeq.empty")
-        }
-      }
-
-      s"""private var $fieldName: $tpeForField = $defaultValue
-         |def $publicName: $publicType = $fieldAccessor""".stripMargin
-    }.mkString("\n\n")
-  }
-
-  def propertyKeyDef(name: String, baseType: String, cardinality: Cardinality) = {
+  def propertyKeyDef(name: String, baseType: String, cardinality: Property.Cardinality) = {
     val completeType = cardinality match {
-      case Cardinality.One       => baseType
-      case Cardinality.ZeroOrOne => baseType
-      case Cardinality.List      => s"Seq[$baseType]"
-      case Cardinality.ISeq=> s"IndexedSeq[$baseType]"
+      case Property.Cardinality.One(_)    => baseType
+      case Property.Cardinality.ZeroOrOne => baseType
+      case Property.Cardinality.List      => s"Seq[$baseType]"
+      case Property.Cardinality.ISeq=> s"IndexedSeq[$baseType]"
     }
     s"""val ${camelCaseCaps(name)} = new overflowdb.PropertyKey[$completeType]("$name") """
   }
 
+  def defaultValueImpl[A](default: Property.Default[A]): String =
+    default.value match {
+      case str: String => s"\"$str\""
+      case char: Char => s"'$char'"
+      case byte: Byte => s"$byte: Byte"
+      case short: Short => s"$short: Short"
+      case int: Int => s"$int: Int"
+      case long: Long => s"$long: Long"
+      case float: Float if float.isNaN => "Float.NaN"
+      case float: Float => s"${float}f"
+      case double: Double if double.isNaN => "Double.NaN"
+      case double: Double => s"${double}d"
+      case other => s"$other"
+    }
+
+  def defaultValueCheckImpl[A](memberName: String, default: Property.Default[A]): String = {
+    val defaultValueSrc = defaultValueImpl(default)
+    default.value match {
+      case float: Float if float.isNaN => s"$memberName.isNaN"
+      case double: Double if double.isNaN => s"$memberName.isNaN"
+      case _ => s"($defaultValueSrc) == $memberName"
```

Comment:
You could possibly simplify this to `$memberName equals $defaultValueSrc`. This does the right thing for floats, and the jvm should get rid of the extra allocation.

On a low level, this would lead to scala boxing the floats and asking jvm for comparison (and this does the right thing). 

I think I like the explicit variant more, though, so don't change it.

I am not entirely sure about permitting NaN in float fields:

JVM comparison of primitive floats is ieee (which is unusable for general programming because of NaN issues), while JVM comparison of boxed floats does the right thing (all NaN compare equal; making only identical NaN compare equal would also be ok for that, but breaking reflexivity is a no-no).  Java / JVM is generally good at handling NaN values, no issues here. 

Scala `==` on the other hand is ieee for floating point values, and the standard library is a tire fire with respect to NaN handling. cf eg https://github.com/scala/bug/issues/11698 for an actual violation of language guarantees, and the general wontfix issue that a `distinct` / `toSet` based on java collections do the right thing, but are incorrect for scala collections (because scala stdlib uses `==`, which is unsafe for generic code).